src/backend/joanie/tests/core/test_api_course_product_relations.py

```diff
@@ -200,12 +190,92 @@ def test_api_course_product_relation_read_detail_anonymous(self):
 
         response = self.client.get(f"/api/v1.0/course-product-relations/{relation.id}/")
 
-        self.assertContains(
-            response,
-            "Authentication credentials were not provided.",
-            status_code=401,
+        self.assertEqual(response.status_code, 400)
+        self.assertEqual(
+            response.json(),
+            {"course": "You must specify a course code to get product details."},
         )
 
+        # But authenticated user with access does
+        user = factories.UserFactory()
+        token = self.generate_token_from_user(user)
+        factories.UserCourseAccessFactory(user=user, course=course)
+        response = self.client.get(
+            f"/api/v1.0/course-product-relations/{relation.id}/",
+            HTTP_AUTHORIZATION=f"Bearer {token}",
+        )
+
+        self.assertEqual(response.status_code, 200)
+        content = response.json()
+        self.assertEqual(content["id"], str(relation.id))
+
+    def test_api_course_product_relation_read_detail_anonymous_with_invalid_course_code(
+        self,
+    ):
+        """
+        Anonymous users should not be able to retrieve a single course product relation
+        if the provided course code does not correspond
+        to the one of the course's relation.
+        """
+        courses = factories.CourseFactory.create_batch(2)
+        product = factories.ProductFactory(type=enums.PRODUCT_TYPE_CREDENTIAL)
+        relation = factories.CourseProductRelationFactory(
+            course=courses[0], product=product
+        )
+
+        response = self.client.get(
+            f"/api/v1.0/course-product-relations/{relation.id}/?course={courses[1].code}"
+        )
+
+        self.assertEqual(response.status_code, 404)
+
+    def test_api_course_product_relation_read_detail_anonymous_with_course_code(self):
+        """
+        Anonymous users should be able to retrieve a single course product relation
+        if a course code is provided.
+        """
+        course = factories.CourseFactory(code="00000")
+        product = factories.ProductFactory(
+            type=enums.PRODUCT_TYPE_CREDENTIAL,
+            target_courses=factories.CourseFactory.create_batch(2),
+        )
+        relation = factories.CourseProductRelationFactory(
+            course=course, product=product
+        )
+
+        with self.assertNumQueries(8):
+            response = self.client.get(
+                f"/api/v1.0/course-product-relations/{relation.id}/?course=00000"
```

Comment:
since the course code is required and is necessary to designate a resource, the urls could be of the form:
`/api/v1.0/course-product-relations/{relation.id}/{course.id}`

Actually since there is unicity of the course/product couple, it could be of the form:
`/api/v1.0/course-product-relations/{course.id}/{product.id}`
So that we can call them directly without having to retrieve the relation id!

and when we call
`/api/v1.0/course-product-relations/{course.id}`
we can get all the products for a course (relations actually)?

Reply:
Actually, The `course_code` is not mandatory at all. I just enforce its use in order to prevent some mistake on the catalog. E.g : in the case an editor adds manually a course run to display a product on the syllabus, if we do not enforce the use of this parameter, we could add a product to the wrong course page without feedback. With this parameter, if the relation does not match the syllabus course code, an error will be raised.

But there are cases where course is not required as retrieve relations on the teacher dashboard.